File: t2m.rb
#!/usr/bin/ruby

require 'socket'
require 'date'
require 'json'
require 'logger'


BUFSIZE = 1024


if test ?d, '/vagrant'
    logfilename = '/vagrant/.t2m.log'
else
    logfilename = '/tmp/.t2m.log'
end
$logger = Logger.new(logfilename)
$logger.level = Logger::DEBUG



unless ARGV.count == 2
  puts "Usage: #{$0} listen_ip port_number"
  exit(1)
end


def s_to_datetime(dts)
    parsed_time = DateTime.strptime(dts, '%d/%m/%Y %H:%M')
    return parsed_time
end


def parse_data(data)
    dtobj = String.new()
    msg = String.new()

    begin
        if data.include? "]"
            dts, msg = data.split(']')
            dts = dts.sub('[', '')
            if msg[0] == ' '
                msg = msg[1..-1]
                dtobj = s_to_datetime(dts)
            end
        end
    ensure
        return dtobj, msg
    end

end


def main()
    listen_ip = ARGV[0]
    port = ARGV[1].to_i

    s = UDPSocket.new
    s.bind(listen_ip, port)

    while true
        mesg, addr = s.recvfrom(BUFSIZE)
        dt, msg = parse_data(mesg)

        if dt != "" and msg != ""
            dict = { "timestamp" => dt.strftime("%s").to_i, "message" => msg }
            print(dict.to_json, "\n")
            $logger.debug("#{dict.to_json}")
            next
        end
        $logger.debug('invalid message received.')
    end
end



main()

File: t2m.py
#! /usr/bin/env python
# Requires python 2.7

import sys
import socket
import datetime
import json
import logging

import os

if os.path.exists('/vagrant'):
    logfilename = '/vagrant/.t2m.log'
else:
    logfilename = '/tmp/.t2m.log'

logging.basicConfig(
    filename=logfilename,
    level=logging.DEBUG)


BUFSIZE = 1024
DEBUG = True


def main():
    if len(sys.argv) < 2:
        usage()
    try:
        server()
    except Exception, e:
        logging.critical("server() exited with exception %s" % str(e))

def usage():
    sys.stdout = sys.stderr
    print('Usage: %s listen_ip udp_port' % sys.argv[0])
    sys.exit(2)


def debug(msg):
    if DEBUG: logging.debug(msg)


def s_to_datetime(dts):
    try:
        dtobj = datetime.datetime.strptime(dts, "%d/%m/%Y %H:%M")
        return dtobj
    except ValueError:
        return False

def parse_data(data):
    """
    Quick n dirty data parsing.
    string -> datetime, string
    """
    dtobj = None
    msg = None
    if ']' not in data:
        return dtobj, msg
    try:
        dts, msg = data.split(']')
        dts = dts.strip('[')
        if msg[0] != ' ':
            return dtobj, msg
        msg = msg[1:]
        dtobj = s_to_datetime(dts)
    except ValueError:
        pass
    return dtobj, msg


def server():
    if len(sys.argv) > 2:
        host = sys.argv[1]
        port = eval(sys.argv[2])

    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.bind((host, port))
    debug('udp server ready on %s:%s' % (host,port))
    while 1:
        data, addr = s.recvfrom(BUFSIZE)
        debug('server received %r from %r' % (data, addr))

        dt, msg = parse_data(data)
        if dt and msg:
            #send data back
            datadict = {
                "timestamp": (dt - datetime.datetime(1970,1,1)).total_seconds(),
                "message": msg
            }
            jsondata = json.dumps(datadict)
            debug(jsondata)
            print(jsondata)
            s.sendto(jsondata, addr)
        else:
            debug('invalid message received.')


if __name__ == '__main__': main()


